src/main/java/com/hyperlocal/server/ShopController.java

```diff
@@ -1,64 +1,233 @@
+
+    // Create the HTTP Request to send
+    HttpClient client = HttpClient.newHttpClient();
+    HttpRequest request = HttpRequest.newBuilder().uri(URI.create(SEARCH_INDEX_URL))
+        .method("GET", HttpRequest.BodyPublishers.ofString(searchSourceBuilder.toString()))
+        .setHeader("Content-Type", "application/json").build();
+
+    // Send request to search Index asynchronously and parse response to get ShopIDs
+    return client.sendAsync(request, BodyHandlers.ofString()).thenApply(HttpResponse::body)
+        .thenApply((responseString) -> {
+          JsonObject obj = JsonParser.parseString(responseString).getAsJsonObject();
+          JsonArray idListJson = obj.get("hits").getAsJsonObject().get("hits").getAsJsonArray();
+          for (JsonElement id : idListJson) {
+            shopIDList.add(id.getAsJsonObject().get("_id").getAsLong());
+          }
+          return shopIDList;
+        
+        // Perform BatchQuery on shopIDs and get List of ShopDetails corresponding to the shopIDs
+        }).thenCompose((shopList) -> {
+          return getShopsByShopIDBatch(shopList);
+        });
+  }
+
+  // API for browsing in a certain radius
+  @GetMapping("/api/browse/elastic")
+  public CompletableFuture<List<ShopDetails>> getDataFromElasticSearch(@RequestParam String queryRadius,
+      @RequestParam String latitude, @RequestParam String longitude) {
+
+    // Prepare query for search Index
+    GeoDistanceQueryBuilder filterOnDistance = QueryBuilders.geoDistanceQuery("pin.location")
+        .point(Double.parseDouble(latitude), Double.parseDouble(longitude)).distance(queryRadius);
+    BoolQueryBuilder boolMatchQueryWithDistanceFilter = QueryBuilders.boolQuery().must(QueryBuilders.matchAllQuery())
+        .filter(filterOnDistance);
+    SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+    searchSourceBuilder.query(boolMatchQueryWithDistanceFilter);
+
+    // Create HTTP Request
+    HttpClient client = HttpClient.newHttpClient();
+    HttpRequest request = HttpRequest.newBuilder().uri(URI.create(SEARCH_INDEX_URL))
+        .method("GET", HttpRequest.BodyPublishers.ofString(searchSourceBuilder.toString()))
+        .setHeader("Content-Type", "application/json").build();
+
+    List<Long> shopIDList = new ArrayList<Long>();
+
+    // Send the request and get list of ShopIDs
+    return client.sendAsync(request, BodyHandlers.ofString()).thenApply(HttpResponse::body)
+        .thenApply((responseString) -> {
+          JsonObject obj = JsonParser.parseString(responseString).getAsJsonObject();
+          JsonArray idListJson = obj.get("hits").getAsJsonObject().get("hits").getAsJsonArray();
+          for (JsonElement id : idListJson) {
+            shopIDList.add(id.getAsJsonObject().get("_id").getAsLong());
+          }
+          return shopIDList;
+      
+        // Get List of ShopDetails for the list of ShopIDs via Batch Query
+        }).thenCompose((shopList) -> {
```

Comment:
In lieu of chaining a `thenCompose()` after the `thenApply()` method, you can get rid of `thenApply()` method altogether and call `getShopsByShopIDBatch()` in the same lambda function which prepares the `shopIDList`.

Reply:
Done. Thanks a lot!!!